### PR TITLE
[java] Deprecate visitor decorators

### LINKED ISSUE #1483: [java] Cyclo metric should count conditions of for statements correctly
E.g.
```java
           class Foo {
                private static int compareImports(String import1, String import2) {
                    for (int i = 0; i != import1Tokens.length && i != import2Tokens.length; i++) {
                    }
                }
            }
```

The current implementation counts the complexity of the expression in the initializer `int i = 0` instead of the boolean expression:
https://github.com/pmd/pmd/blob/1db950bec62df857ae734c1b01ccc0e1876e7124/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/CycloPathAwareDecorator.java#L49

### PR DESCRIPTION
**Pull request by oowekyala**
* They're a pain to maintain, and are very much overengineered.  
* When we change parts of the grammar with 7.0.0, they're two additionnal duplications of JavaParserVisitorAdapter.
* Removing them required a rewrite of the visitors for Cyclo and Ncss, but the tests didn't move, and it's probably a not-so-used part of PMD, so I think it's safe to say it's low-impact

Closes #1483

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 23 new violations and 0 new errors,<br />
removes 27 violations and 0 errors. <a href="http://chunk.io/pmd/c620840cef934990a24ad038fc4f3b39/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTIfStatement.java [hunk 1]
```diff
@@ -5,29 +5,79 @@
...
+    public ASTExpression getGuardExpressionNode() {
+        return (ASTExpression) jjtGetChild(0);
+    }
+
+
+    /**
+     * Returns the statement that will be run if the guard evaluates
+     * to true.
+     */
+    public ASTStatement getTrueAlternative() {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTIfStatement.java]: I'm unsure about the names getTrueAlternative / getFalseAlternative. In Rhino (Javascript), they have named it "getThenPart" and "getElsePart". In JLS, they talk about statements (first statement and second statement).
Not so helpfull, either.

How about these names?

* `getCondition()`
* `getThenBranch()`
* `getElseBranch()`

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/CycloVisitor.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import net.sourceforge.pmd.lang.metrics.MetricOptions;
+
+
+/**
+ * Visitor for the Cyclo metric.
+ *
+ * @author Clément Fournier
+ * @since 6.7.0
+ */
+public class CycloVisitor extends JavaParserVisitorAdapter {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/CycloVisitor.java]: I guess, this class itself is internal (it's inside a subpackage of a package called `impl`).
Should we annotate this then with `@InternalApi` as well?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/NcssVisitor.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+import net.sourceforge.pmd.lang.metrics.MetricOptions;
+
+
+/**
+ * Visitor for the Ncss metric.
+ *
+ * @author Clément Fournier
+ * @since 6.7.0
+ */
+public class NcssVisitor extends JavaParserVisitorAdapter {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/NcssVisitor.java]: InternalApi?

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml [hunk 4]
```diff
@@ -394,4 +394,27 @@
...
     </test-code>
 
+    <!-- Guard of for statements were previously wrongly counted. -->
+    <!-- See https://github.com/pmd/pmd/blob/0c1afc426debff979b25219dbc8f4e07c76645b5/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/CycloPathAwareDecorator.java#L53 -->
+    <!-- The method fetches the first descendant as the guard expression, though it should be the first child -->
+    <!-- This causes the method to fetch the initializer of 'i' in the following for statement ('0') -->
+    <!-- instead of the more complex boolean expression. -->
+
+    <test-code>
+        <description>The guard of a ForStatement should be counted correctly (spotted #1437)</description>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml]: Is this bugfix worth an own issue? Then we can mention it in the release notes...

File: docs/pages/release_notes.md [hunk 5]
```diff
@@ -22,6 +22,17 @@ This is a {{ site.pmd.release_type }} release.
...
+#### Deprecated APIs
+
+* Visitor decorators are now deprecated and will be removed in PMD 7.0.0. They were originally a way to write
+  composable visitors, used in the metrics framework, but they didn't prove cost-effective.
+  * In `net.sourceforge.pmd.lang.java.ast`: JavaParserDecoratedVisitor, JavaParserControllessVisitor,
+    JavaParserControllessVisitorAdapter, and JavaParserVisitorDecorator are deprecated with no intended replacement
+  * In `net.sourceforge.pmd.lang.java.metrics.impl`:
+    * CycloAssertAwareDecorator, CycloBaseVisitor, and CycloPathAwareDecorator are deprecated and are replaced with CycloVisitor
+    * NcssBaseVisitor and NcssCountImportsDecorator are deprecated and are replaced with NcssVisitor
+
```
Comment (adangel) [inline docs/pages/release_notes.md]: We should propose an alternative, e.g. "For implementing own visitors, a JavaParserVisitorAdapter should be subclassed directly".

**Comment by oowekyala** [comment]
TODO make the contents of the metrics.impl packages private API (deprecate, update doc)

**Comment by oowekyala** [comment]
@adangel I tagged the contents of `metrics.impl.visitors` as internal API, but not the actual metrics in `metrics.impl`.Those e.g. declare MetricOptions in that package, and are documented there, so they're part of our API.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml (hunk 4 above)
Comment (oowekyala) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml]: I created #1483 for that

**Comment by oowekyala** [comment]
Closes #1483

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTIfStatement.java (hunk 1 above)
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTIfStatement.java]: I renamed the methods to get the branches. I'm not sure about `getCondition`. It's certainly a better name, but all other statement nodes have their condition named `getGuardExpressionNode`. Maybe we should rename them all to `getCondition` ?

Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTIfStatement.java]: Maybe, not sure. If we rename it, then in a different PR and we should use the same name in all the nodes. For now, it is consistent as it is.

**Pull request merged**